feat(models,training): Ensemble Interpolator

Magnus-SI opened the pull request:
## Description
Allows for training a probabilistic interpolator/temporal downscaler by using the same method as the ensemble forecaster.

##  Additional notes ##
This branch and draft PR is for testing purposes. Cleaner code and proper tests will be added once we can confirm that this method works well. But feel free to comment on the method itself.

## TODO ##
- [ ] Implement target parallelism for faster training.
- [x] Complete a full training run and verify.
- [x] Consider adding the option to output all targets simultaneously.

jakob-schloer committed:
feat: Callbacks for GraphEnsForecaster. (#449)

## Description
Callbacks for GraphEnsForecaster.

Includes the following callbacks now:

```
callbacks:
  - _target_: anemoi.training.diagnostics.callbacks.evaluation.RolloutEvalEns
    rollout: ${dataloader.validation_rollout}
    every_n_batches: 4

plot:
  callbacks:
    # Add plot callbacks here.
    - _target_: anemoi.training.diagnostics.callbacks.plot_ens.PlotEnsSample
      sample_idx: ${diagnostics.plot.sample_idx}
      per_sample : 2
      parameters: ${diagnostics.plot.parameters}
      every_n_batches: ${diagnostics.plot.frequency.batch}
      #Defining the accumulation levels for precipitation related fields and the colormap
      accumulation_levels_plot: [0, 0.05, 0.1, 0.25, 0.5, 1, 1.5, 2, 3, 4, 5, 6, 7, 100] # in mm

    # Deterministic callbacks are also overloaded.
    - _target_: anemoi.training.diagnostics.callbacks.plot_ens.PlotLoss
      # group parameters by categories when visualizing contributions to the loss
      # one-parameter groups are possible to highlight individual parameters
      parameter_groups:
        moisture: [tp, cp, tcw]
        sfc_wind: [10u, 10v]
      every_n_batches: ${diagnostics.plot.frequency.batch}
    - _target_: anemoi.training.diagnostics.callbacks.plot_ens.PlotSpectrum
      sample_idx: ${diagnostics.plot.sample_idx}
      parameters: ${diagnostics.plot.parameters}
      every_n_batches: ${diagnostics.plot.frequency.batch}
    - _target_: anemoi.training.diagnostics.callbacks.plot_ens.PlotHistogram
      sample_idx: ${diagnostics.plot.sample_idx}
      parameters: ${diagnostics.plot.parameters}
      every_n_batches: ${diagnostics.plot.frequency.batch}
      precip_and_related_fields: ["tp", "cp"] # Optional: specify precip fields for special histogram treatment
    - _target_: anemoi.training.diagnostics.callbacks.plot_ens.GraphTrainableFeaturesPlot
      every_n_epochs: ${diagnostics.plot.frequency.epoch}

    # Overloaded PlotSample will return the plots for the first ensemble member
    # - _target_: anemoi.training.diagnostics.callbacks.plot_ens.PlotSample
    #   sample_idx: ${diagnostics.plot.sample_idx}
    #   per_sample: 6
    #   parameters: ${diagnostics.plot.parameters}
    #   every_n_batches: ${diagnostics.plot.frequency.batch}
    #   accumulation_levels_plot: [0, 0.05, 0.1, 0.25, 0.5, 1, 1.5, 2, 3, 4, 5, 6, 7, 100] # in mm
    #   precip_and_related_fields: ["tp", "cp"] # Optional: specify precip fields
```

## Potential future work
In `aifs-mono` we still had two more callbacks which we did not bring
over:
- `RankHistogramPlots`
- `SpreadSkill metrics`
I would suggest making separate PRs for that.

## Testing

Here are MLFlow test runs with the callbacks: 
- deterministic:
https://mlflow.ecmwf.int/#/experiments/395/runs/77f389f63b964cd596e7f7554b2662ac
- ensemble:
https://mlflow.ecmwf.int/#/experiments/266/runs/06dca3fea6a1494fa4ab88eebcd833e2

Integration test:
https://github.com/ecmwf/anemoi-core/actions/runs/18715469239
Additional: 
- included fix for the broken LongRollout callback

---------

Co-authored-by: Mariana Clare <mariana.clare@ecmwf.int>
Co-authored-by: Simon Lang <simon.lang@ecmwf.int>
Co-authored-by: sahahner <hahner.sa@gmail.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Mariana Clare <31656450+mc4117@users.noreply.github.com>

jakob-schloer committed:
fix: Bug in sample plots. (#632)

## Description
Fix a bug in the PlotSample callbacks when `validation_rollout=1`. 

## Example run
Deterministic:
https://mlflow.ecmwf.int/#/experiments/395/runs/5f1adb3b789147a1a0ab905671301b58
Ensemble:
https://mlflow.ecmwf.int/#/experiments/413/runs/13c76c9adc534bd08cff25d3ec520585

## Nightly integration test
https://github.com/ecmwf/anemoi-core/actions/runs/18880348799

VeraChristina committed:
chore: remove duplicate plotting code (#639)

## Description
Removing duplicate plotting code, see lines 936-968


***As a contributor to the Anemoi framework, please ensure that your
changes include unit tests, updates to any affected dependencies and
documentation, and have been tested in a parallel setting (i.e., with
multiple GPUs). As a reviewer, you are also responsible for verifying
these aspects and requesting changes if they are not adequately
addressed. For guidelines about those please refer to
https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the
[Contributor License
Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md)

Magnus-SI committed:
feat(training,models): add ensemble interpolator prototype

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

mc4117 commented on training/src/anemoi/training/train/tasks/ensinterp.py:
```diff
@@ -0,0 +1,395 @@
+LOGGER = logging.getLogger(__name__)
+
+
+class GraphEnsInterp(BaseGraphModule):
```
same here? It would remove a lot of need for duplicate code and make it easier to see what are the real changes needed for EnsInterpolators?

Magnus-SI replied:
Agreed, though I was planning to inherit from the GraphEnsForecaster class here. The GraphInterpolator class only modifies __init__ _step and forward, and we need to change __init__ and _step anyway. forward would be the only function inheritable as is. While from the GraphEnsForecaster, everything related to ensemble parallel is fully inheritable.

Magnus-SI committed:
feat: option for one interpolator target per iteration

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Magnus-SI committed:
add example config

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Magnus-SI committed:
add experimental multi-target ensinterp

anaprietonem committed:
feat: activate minmium plotting for integration tests (#669)

Testing plots it's not part our current testing suite. Since most of
them are implemented via callbacks a simple way would be to activate a
minimal plotting set up for the integration test to ensure we don't
break them in the future.

This PR proposes those changes:
- Spectrum plot it's quite slow so for now it's deactivated for the SG
model
- Since the test dataset for LAM and SG is a small one we just test
plotting tp
- The plots for the interpolator are deactivated - while the callbacks
run the output produced is not scientifically correct. More work will be
done in the future to address this (@mc4117 - notified about this in
slack channel)
- In the hierarchical model - current plotting of edge features is not
support - this plot is for now deactivated. (Discussed with @icedoom888
)

cc: @VeraChristina - having those tests would also be useful in the
context of multiple datasets to ensure there is no feature regression
for plotting.

<!-- Describe if it's a bugfix, new feature, doc update, or breaking
change -->

<!-- link to Issue Number -->

<!-- Include any additional information, caveats, or considerations that
the reviewer should be aware of. -->

***As a contributor to the Anemoi framework, please ensure that your
changes include unit tests, updates to any affected dependencies and
documentation, and have been tested in a parallel setting (i.e., with
multiple GPUs). As a reviewer, you are also responsible for verifying
these aspects and requesting changes if they are not adequately
addressed. For guidelines about those please refer to
https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the
[Contributor License
Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md)

---------

Co-authored-by: Mariana Clare <mariana.clare@ecmwf.int>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

mc4117 committed:
feat: time_interpolator_callbacks (#677)

## Description
Introduce deterministic and ensemble callback plots for the time interpolator. These
were previously broken because they only showed the first step of the
time interpolator

***As a contributor to the Anemoi framework, please ensure that your
changes include unit tests, updates to any affected dependencies and
documentation, and have been tested in a parallel setting (i.e., with
multiple GPUs). As a reviewer, you are also responsible for verifying
these aspects and requesting changes if they are not adequately
addressed. For guidelines about those please refer to
https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the
[Contributor License
Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md)

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Ana Prieto Nemesio <91897203+anaprietonem@users.noreply.github.com>

Magnus-SI committed:
feat: time aggregated scores

mc4117 committed:
fix to batch (#732)

## Description
Fix to initial conditions in ensemble time interpolator

Previously I think the code was using a single initial condition

***As a contributor to the Anemoi framework, please ensure that your
changes include unit tests, updates to any affected dependencies and
documentation, and have been tested in a parallel setting (i.e., with
multiple GPUs). As a reviewer, you are also responsible for verifying
these aspects and requesting changes if they are not adequately
addressed. For guidelines about those please refer to
https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the
[Contributor License
Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md)

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

Magnus-SI committed:
feat: added diff to time-aggregated scores, updated default config

mc4117 commented:
This has been implemented instead by PR #896 and PR #1069
